src/toil/lib/toilDocker.py

```diff
@@ -0,0 +1,213 @@
+from docker.api.container import ContainerApiMixin
+
+from bd2k.util.retry import retry
+from docker import client
+from pwd import getpwuid
+
+from toil.lib import dockerPredicate
+from toil.lib import FORGO
+from toil.lib import STOP
+from toil.lib import RM
+from toil.test import timeLimit
+
+_logger = logging.getLogger(__name__)
+
+
+def dockerCheckOutput(job, tool, parameters=None, workDir=None, dockerParameters=None, deferParam=None,
+                      checkOutput=True, containerName=None):
+    return dockerCall(job=job, tool=tool, parameters=parameters, workDir=workDir, dockerParameters=dockerParameters,
+                      deferParam=deferParam, checkOutput=True, containerName=containerName)
+
+def dockerCall(job, tool, parameters=[], workDir=None,
+               dockerParameters=None, deferParam=None, checkOutput=False, containerName=None, outfile=None,
+               entrypoint=['/bin/bash', '-c'], detach=False, stderr=None, dockerAPIparams=None, data_dir='/data',
+               logDriver=None, removeUponExit=False):
+    """
+    Throws CalledProcessorError if the Docker invocation returns a non-zero exit code
+    This function blocks until the subprocess call to Docker returns
+    :param toil.Job.job job: The Job instance for the calling function.
+    :param str tool: Name of the Docker image to be used (e.g. quay.io/ucsc_cgl/samtools:latest).
+    :param list[str] parameters: Command line arguments to be passed to the tool.
+           If list of lists: list[list[str]], then treat as successive commands chained with pipe.
+    :param str workDir: Directory to mount into the container via `-v`. Destination convention is /data
+    :param list[str] dockerParameters: Parameters to pass to Docker. Default parameters are `--rm`,
+            `--log-driver none`, and the mountpoint `-v work_dir:/data` where /data is the destination convention.
+             These defaults are removed if docker_parmaters is passed, so be sure to pass them if they are desired.
+    :param file outfile: Pipe output of Docker call to file handle
+    :param int deferParam: What action should be taken on the container upon job completion?
+           FORGO (0) will leave the container untouched.
+           STOP (1) will attempt to stop the container with `docker stop` (useful for debugging).
+           RM (2) will stop the container and then forcefully remove it from the system
+           using `docker rm -f`. This is the default behavior if defer is set to None.
+    """
+    thisUser = os.getuid()
+    thisGroup = os.getgid()
+
+    if containerName is None:
+        containerName = _getContainerName(job)
+    if workDir is None:
+        workDir = os.getcwd()
+
+    # If 'parameters' is a list of lists, treat each list as a separate command and
+    # chain with pipes.
+    if len(parameters) > 0 and type(parameters[0]) is list:
+        chain_params = [' '.join(p) for p in [map(pipes.quote, q) for q in parameters]]
```

Comment:
I don't like this comprehension; it's mixing comprehensions and plain `map`, and it's creating a whole intermediate list.

I think it wants to be something like:

```
chain_params = [' '.join((pipes.quote(arg) for arg in command)) for command in parameters]
```

Reply:
This was the same comprehension used before in the old code so I didn't think it was worth changing.  I'll update it.

Reply:
Updated.